### PR TITLE
DT-479: Track failure to login later in process to prevent duplicates

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/MissingCredentialsException.java [hunk 1]
```diff
@@ -1,8 +1,8 @@
 package uk.gov.justice.digital.hmpps.oauth2server.security;
 
-import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.authentication.AccountStatusException;
 
-class MissingCredentialsException extends AuthenticationException {
+class MissingCredentialsException extends AccountStatusException {
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/MissingCredentialsException.java]: account status exception will prevent us retrying all authentication providers.  In the missing credentials case there is absolutely no point trying all three.

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserStateAuthenticationFailureHandler.kt [hunk 2]
```diff
@@ -57,21 +61,36 @@ class UserStateAuthenticationFailureHandler(private val tokenService: TokenServi
...
+            Pair("missinguser", null)
+          }
+        } else {
+          Pair("missingpass", null)
         }
       }
-      is DeliusAuthenticationServiceException -> builder.add("invalid").add("deliusdown")
-      else -> builder.add("invalid")
+      is DeliusAuthenticationServiceException -> Pair("invalid", "deliusdown")
+      else -> Pair("invalid", null)
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserStateAuthenticationFailureHandler.kt]: That's a few nulls... better to do listOf?

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaControllerTest.kt [hunk 3]
```diff
@@ -109,6 +109,15 @@ class MfaControllerTest {
     assertThat(modelAndView).isNull()
   }
 
+  @Test
+  fun `mfaChallenge success telemetry`() {
+    val user = User.of("someuser")
+    whenever(tokenService.getToken(any(), anyString())).thenReturn(Optional.of(UserToken("otken", TokenType.MFA, null, user)))
```
Comment (mikehalmamoj) [inline src/test/java/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaControllerTest.kt]: typo

Comment (petergphillips) [reply src/test/java/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaControllerTest.kt]: deliberate - I try not to create inputs with the same name as the parameter :grin:

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserStateAuthenticationFailureHandler.kt (hunk 2 above)
Comment (petergphillips) [reply src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserStateAuthenticationFailureHandler.kt]: Not sure.  I initially had it as a `listOf` but had a problem in that it should really be two parameters - one required and the other optional.  I therefore had to do a `first` with a throw and looked a bit worse - I'll raise another PR to explain better

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaControllerTest.kt (hunk 3 above)
Comment (mikehalmamoj) [reply src/test/java/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaControllerTest.kt]: Ah, interesting.  On my last project we got quite anal about the intent of test data - "some token" would indicate a value to be tested and "any token" would indicate that we don't care about the value in this test.  Which accidentally makes them a different name to the parameter.

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserStateAuthenticationFailureHandler.kt (hunk 2 above)
Comment (mikehalmamoj) [reply src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserStateAuthenticationFailureHandler.kt]: I don't really have a strong opinion either way, just wondered about your rationale.

**Pull request merged**